tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -937,31 +975,86 @@ def _generate_build_extra_metadata_for_tests(tests, bazel_rules):
     'deps("//src/proto/...")',
 ]
 
+# Step 1: run a bunch of "bazel query --output xml" queries to collect
+# the raw build metadata from the bazel build.
+# At the end of this step we will have a dictionary of bazel rules
+# that are interesting to us (libraries, binaries, etc.) along
+# with their most important metadata (sources, headers, dependencies)
 bazel_rules = {}
 for query in _BAZEL_DEPS_QUERIES:
     bazel_rules.update(
         _extract_rules_from_bazel_xml(_bazel_query_xml_tree(query)))
 
+# Step 1a: Knowing the transitive closure of dependencies will make
+# the postprocessing simpler, so compute the info for all our rules.
 _populate_transitive_deps(bazel_rules)
 
-tests = _filter_cc_tests(_extract_cc_tests(bazel_rules))
-test_metadata = _generate_build_extra_metadata_for_tests(tests, bazel_rules)
-
-all_metadata = {}
-all_metadata.update(_BUILD_EXTRA_METADATA)
-all_metadata.update(test_metadata)
-
-all_targets_dict = _generate_build_metadata(all_metadata, bazel_rules)
+# Step 2: Extract the known bazel cc_test tests. While most tests
+# will be buildable with other build systems just fine, some of these tests
+# would be too difficult to build and run with other build systems,
+# so we simply the ones we don't want.
+tests = _exclude_unwanted_cc_tests(_extract_cc_tests(bazel_rules))
+
+# Step 3: Generate the "extra metadata" for all our build targets.
+# While the bazel rules give us most of the information we need,
+# the legacy "build.yaml" format requires some additional fields that
```

Comment:
Can you list the specific extra fields that need to be populated and what each of them indicates?

Reply:
The fields being set and explanation for some of them is here:
https://github.com/grpc/grpc/blob/9e095b2076e1df72d2241128b9c6714d4cc8264b/tools/buildgen/extract_metadata_from_bazel_xml.py#L575

General documentation of "build.yaml" contents is here:
https://github.com/grpc/grpc/blob/master/templates/README.md
That is good enough for general understanding what the fields mean, but the docs might be out of date or incorrect in some cases (build.yaml has been with us from pretty much the project start). While some of the field name seem self-explanatory, I don't really have a way to know what exactly the fields are supposed to mean because they way the they were added to `build.yaml` was mostly random and they were added on ad-hoc and as-needed fashion. So when someone needed a new field to achieve something in our build system, they added a new attribute and the exact meaning was defined implicitly by implementing its behavior in our build file generators.

For the above reasons, there's actually no clearly defined list of extra fields that need to be populated - it's basically empirical (I basically just tried to match exactly what was in the original build.yaml because that was the only way to prevent things from exploding - there are too many special cases in our build generators to address all of them at once).

In the future, the right thing to do is to revisit the "extra metadata" fields being used in _BUILD_EXTRA_METADATA and get rid of all of them one by one  (either by removing the need for a field entirely or somehow adding the information to the bazel rule so we can consume it) - unlike before, this is now possible because bazel build is now our source of truth and the big bazel related changes have already settled down (and most builds are mostly green).